plugins/source/azure/client/client.go

```diff
@@ -65,49 +67,89 @@ func (c *Client) discoverSubscriptions(ctx context.Context) error {
+	providerClient, err := armresources.NewProvidersClient(subscriptionId, c.Creds, c.Options)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create provider client: %w", err)
+	}
+	providers := make([]*armresources.Provider, 0)
+	providerPager := providerClient.NewListPager(nil)
+	for providerPager.More() {
+		providerPage, err := providerPager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to list providers: %w", err)
+		}
+		if len(providerPage.Value) == 0 {
+			continue
+		}
+		for _, p := range providerPage.Value {
+			if p.RegistrationState != nil && *p.RegistrationState == "Registered" {
+				providers = append(providers, p)
+			}
+		}
+	}
+	return providers, nil
+}
+
 func (c *Client) discoverResourceGroups(ctx context.Context) error {
 	c.ResourceGroups = make(map[string][]*armresources.ResourceGroup, len(c.subscriptions))
 	c.registeredNamespaces = make(map[string]map[string]bool, len(c.subscriptions))
 
+	groupsLock, namespacesLock := sync.Mutex{}, sync.Mutex{}
+
+	errorGroup, gtx := errgroup.WithContext(ctx)
 	for _, subID := range c.subscriptions {
-		c.registeredNamespaces[subID] = make(map[string]bool)
-		cl, err := armresources.NewResourceGroupsClient(subID, c.Creds, c.Options)
-		if err != nil {
-			return fmt.Errorf("failed to create resource group client: %w", err)
-		}
-		pager := cl.NewListPager(&armresources.ResourceGroupsClientListOptions{})
-		for pager.More() {
-			page, err := pager.NextPage(ctx)
+		subID := subID
+		errorGroup.Go(func() error {
+			groups, err := c.getResourceGroupsForSubscription(gtx, subID)
 			if err != nil {
-				return fmt.Errorf("failed to list resource groups: %w", err)
+				return err
 			}
-			if len(page.Value) == 0 {
-				continue
-			}
-			c.ResourceGroups[subID] = append(c.ResourceGroups[subID], page.Value...)
-		}
+			groupsLock.Lock()
+			c.ResourceGroups[subID] = groups
+			groupsLock.Unlock()
 
-		providerClient, err := armresources.NewProvidersClient(subID, c.Creds, c.Options)
-		if err != nil {
-			return fmt.Errorf("failed to create provider client: %w", err)
-		}
-		providerPager := providerClient.NewListPager(nil)
-		for providerPager.More() {
-			providerPage, err := providerPager.NextPage(ctx)
+			return nil
+		})
+
+		errorGroup.Go(func() error {
+			providers, err := c.getRegisteredProvidersForSubscription(gtx, subID)
 			if err != nil {
-				return fmt.Errorf("failed to list providers: %w", err)
-			}
-			if len(providerPage.Value) == 0 {
-				continue
+				return err
 			}
-			for _, p := range providerPage.Value {
-				if p.RegistrationState != nil && *p.RegistrationState == "Registered" {
-					c.registeredNamespaces[subID][strings.ToLower(*p.Namespace)] = true
-				}
+
+			namespacesLock.Lock()
+			c.registeredNamespaces[subID] = make(map[string]bool)
+			for _, p := range providers {
+				c.registeredNamespaces[subID][strings.ToLower(*p.Namespace)] = true
 			}
-		}
+			namespacesLock.Unlock()
```

Comment:
I would move this to a defer statement just after the `.Lock()` call, to make sure it runs even after a panic